{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from assocplots.qqplot import *\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_df = pd.read_csv('/storage/cynthiawu/trans_eQTL/GTex.v8/Nerve_Tibial/gene-snp_eqtls_empiricalpvalues_allchr', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>cpma</th>\n",
       "      <th>pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_100007048</td>\n",
       "      <td>175.076413</td>\n",
       "      <td>0.722275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_100110377</td>\n",
       "      <td>179.462116</td>\n",
       "      <td>0.718867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_100133310</td>\n",
       "      <td>240.862862</td>\n",
       "      <td>0.676731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_100133311</td>\n",
       "      <td>240.862862</td>\n",
       "      <td>0.676731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_100206504</td>\n",
       "      <td>392.056652</td>\n",
       "      <td>0.593199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31961</th>\n",
       "      <td>chr22_50606762</td>\n",
       "      <td>138.655291</td>\n",
       "      <td>0.776340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31962</th>\n",
       "      <td>chr22_50625611</td>\n",
       "      <td>211.999535</td>\n",
       "      <td>0.725075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31963</th>\n",
       "      <td>chr22_50626933</td>\n",
       "      <td>294.377515</td>\n",
       "      <td>0.678135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31964</th>\n",
       "      <td>chr22_50679152</td>\n",
       "      <td>56.247171</td>\n",
       "      <td>0.856474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31965</th>\n",
       "      <td>chr22_50744827</td>\n",
       "      <td>56.022188</td>\n",
       "      <td>0.856760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31966 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp        cpma    pvalue\n",
       "0      chr1_100007048  175.076413  0.722275\n",
       "1      chr1_100110377  179.462116  0.718867\n",
       "2      chr1_100133310  240.862862  0.676731\n",
       "3      chr1_100133311  240.862862  0.676731\n",
       "4      chr1_100206504  392.056652  0.593199\n",
       "...               ...         ...       ...\n",
       "31961  chr22_50606762  138.655291  0.776340\n",
       "31962  chr22_50625611  211.999535  0.725075\n",
       "31963  chr22_50626933  294.377515  0.678135\n",
       "31964  chr22_50679152   56.247171  0.856474\n",
       "31965  chr22_50744827   56.022188  0.856760\n",
       "\n",
       "[31966 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "conseq_path = '/storage/cynthiawu/trans_eQTL/GTex.v8/VariantConseq/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_variant_effect.pkl'\n",
    "with open(conseq_path, 'rb') as f:\n",
    "    conseq_dict = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1:10144': 'downstream_gene_variant',\n",
       " 'chr1:10177': 'downstream_gene_variant',\n",
       " 'chr1:10180': 'downstream_gene_variant',\n",
       " 'chr1:10181': 'downstream_gene_variant',\n",
       " 'chr1:10228': 'downstream_gene_variant',\n",
       " 'chr1:10250': 'downstream_gene_variant',\n",
       " 'chr1:10257': 'downstream_gene_variant',\n",
       " 'chr1:10291': 'downstream_gene_variant',\n",
       " 'chr1:10297': 'downstream_gene_variant',\n",
       " 'chr1:10303': 'downstream_gene_variant',\n",
       " 'chr1:10321': 'downstream_gene_variant',\n",
       " 'chr1:10327': 'downstream_gene_variant',\n",
       " 'chr1:10328': 'downstream_gene_variant',\n",
       " 'chr1:10329': 'downstream_gene_variant',\n",
       " 'chr1:10330': 'downstream_gene_variant',\n",
       " 'chr1:10334': 'downstream_gene_variant',\n",
       " 'chr1:10340': 'downstream_gene_variant',\n",
       " 'chr1:10346': 'downstream_gene_variant',\n",
       " 'chr1:10352': 'downstream_gene_variant',\n",
       " 'chr1:10354': 'downstream_gene_variant',\n",
       " 'chr1:10419': 'downstream_gene_variant',\n",
       " 'chr1:10423': 'downstream_gene_variant',\n",
       " 'chr1:10432': 'downstream_gene_variant',\n",
       " 'chr1:10437': 'downstream_gene_variant',\n",
       " 'chr1:10440': 'downstream_gene_variant',\n",
       " 'chr1:10443': 'downstream_gene_variant',\n",
       " 'chr1:10444': 'downstream_gene_variant',\n",
       " 'chr1:10446': 'downstream_gene_variant',\n",
       " 'chr1:10450': 'downstream_gene_variant',\n",
       " 'chr1:10452': 'downstream_gene_variant',\n",
       " 'chr1:10455': 'downstream_gene_variant',\n",
       " 'chr1:10456': 'downstream_gene_variant',\n",
       " 'chr1:10458': 'downstream_gene_variant',\n",
       " 'chr1:10462': 'downstream_gene_variant',\n",
       " 'chr1:10469': 'downstream_gene_variant',\n",
       " 'chr1:10470': 'downstream_gene_variant',\n",
       " 'chr1:10473': 'downstream_gene_variant',\n",
       " 'chr1:10474': 'downstream_gene_variant',\n",
       " 'chr1:10480': 'downstream_gene_variant',\n",
       " 'chr1:10482': 'downstream_gene_variant',\n",
       " 'chr1:10492': 'downstream_gene_variant',\n",
       " 'chr1:10579': 'downstream_gene_variant',\n",
       " 'chr1:10583': 'downstream_gene_variant',\n",
       " 'chr1:10610': 'downstream_gene_variant',\n",
       " 'chr1:10611': 'downstream_gene_variant',\n",
       " 'chr1:10622': 'downstream_gene_variant',\n",
       " 'chr1:10623': 'downstream_gene_variant',\n",
       " 'chr1:10719': 'downstream_gene_variant',\n",
       " 'chr1:10774': 'downstream_gene_variant',\n",
       " 'chr1:10904': 'downstream_gene_variant',\n",
       " 'chr1:10914': 'downstream_gene_variant',\n",
       " 'chr1:10927': 'downstream_gene_variant',\n",
       " 'chr1:10938': 'downstream_gene_variant',\n",
       " 'chr1:10947': 'downstream_gene_variant',\n",
       " 'chr1:10984': 'downstream_gene_variant',\n",
       " 'chr1:10990': 'downstream_gene_variant',\n",
       " 'chr1:11014': 'downstream_gene_variant',\n",
       " 'chr1:11022': 'downstream_gene_variant',\n",
       " 'chr1:12659': 'downstream_gene_variant',\n",
       " 'chr1:12672': 'downstream_gene_variant',\n",
       " 'chr1:12693': 'downstream_gene_variant',\n",
       " 'chr1:12719': 'downstream_gene_variant',\n",
       " 'chr1:12755': 'downstream_gene_variant',\n",
       " 'chr1:12774': 'downstream_gene_variant',\n",
       " 'chr1:12783': 'downstream_gene_variant',\n",
       " 'chr1:12807': 'downstream_gene_variant',\n",
       " 'chr1:12829': 'downstream_gene_variant',\n",
       " 'chr1:12832': 'downstream_gene_variant',\n",
       " 'chr1:12868': 'downstream_gene_variant',\n",
       " 'chr1:12888': 'downstream_gene_variant',\n",
       " 'chr1:12894': 'downstream_gene_variant',\n",
       " 'chr1:12901': 'downstream_gene_variant',\n",
       " 'chr1:12948': 'downstream_gene_variant',\n",
       " 'chr1:13007': 'downstream_gene_variant',\n",
       " 'chr1:13011': 'downstream_gene_variant',\n",
       " 'chr1:13012': 'downstream_gene_variant',\n",
       " 'chr1:13065': 'downstream_gene_variant',\n",
       " 'chr1:13074': 'downstream_gene_variant',\n",
       " 'chr1:13079': 'downstream_gene_variant',\n",
       " 'chr1:13244': 'downstream_gene_variant',\n",
       " 'chr1:13248': 'downstream_gene_variant',\n",
       " 'chr1:13268': 'downstream_gene_variant',\n",
       " 'chr1:13270': 'downstream_gene_variant',\n",
       " 'chr1:13271': 'downstream_gene_variant',\n",
       " 'chr1:13273': 'downstream_gene_variant',\n",
       " 'chr1:13302': 'downstream_gene_variant',\n",
       " 'chr1:13380': 'downstream_gene_variant',\n",
       " 'chr1:13415': 'downstream_gene_variant',\n",
       " 'chr1:13418': 'downstream_gene_variant',\n",
       " 'chr1:13494': 'downstream_gene_variant',\n",
       " 'chr1:13504': 'downstream_gene_variant',\n",
       " 'chr1:13508': 'downstream_gene_variant',\n",
       " 'chr1:13526': 'downstream_gene_variant',\n",
       " 'chr1:13528': 'downstream_gene_variant',\n",
       " 'chr1:13529': 'downstream_gene_variant',\n",
       " 'chr1:13537': 'downstream_gene_variant',\n",
       " 'chr1:13550': 'downstream_gene_variant',\n",
       " 'chr1:13595': 'downstream_gene_variant',\n",
       " 'chr1:13613': 'downstream_gene_variant',\n",
       " 'chr1:13645': 'downstream_gene_variant',\n",
       " 'chr1:13649': 'downstream_gene_variant',\n",
       " 'chr1:13668': 'downstream_gene_variant',\n",
       " 'chr1:13684': 'downstream_gene_variant',\n",
       " 'chr1:13750': 'downstream_gene_variant',\n",
       " 'chr1:13751': 'downstream_gene_variant',\n",
       " 'chr1:13752': 'downstream_gene_variant',\n",
       " 'chr1:13757': 'downstream_gene_variant',\n",
       " 'chr1:13760': 'downstream_gene_variant',\n",
       " 'chr1:13761': 'downstream_gene_variant',\n",
       " 'chr1:13770': 'downstream_gene_variant',\n",
       " 'chr1:13774': 'downstream_gene_variant',\n",
       " 'chr1:13777': 'downstream_gene_variant',\n",
       " 'chr1:13788': 'downstream_gene_variant',\n",
       " 'chr1:13811': 'downstream_gene_variant',\n",
       " 'chr1:13813': 'downstream_gene_variant',\n",
       " 'chr1:13824': 'downstream_gene_variant',\n",
       " 'chr1:13835': 'downstream_gene_variant',\n",
       " 'chr1:13838': 'downstream_gene_variant',\n",
       " 'chr1:13862': 'downstream_gene_variant',\n",
       " 'chr1:13868': 'downstream_gene_variant',\n",
       " 'chr1:13912': 'downstream_gene_variant',\n",
       " 'chr1:13949': 'downstream_gene_variant',\n",
       " 'chr1:13980': 'downstream_gene_variant',\n",
       " 'chr1:14026': 'downstream_gene_variant',\n",
       " 'chr1:14115': 'downstream_gene_variant',\n",
       " 'chr1:14116': 'downstream_gene_variant',\n",
       " 'chr1:14133': 'downstream_gene_variant',\n",
       " 'chr1:14165': 'downstream_gene_variant',\n",
       " 'chr1:14170': 'downstream_gene_variant',\n",
       " 'chr1:14180': 'downstream_gene_variant',\n",
       " 'chr1:14210': 'downstream_gene_variant',\n",
       " 'chr1:14224': 'downstream_gene_variant',\n",
       " 'chr1:14248': 'downstream_gene_variant',\n",
       " 'chr1:14264': 'downstream_gene_variant',\n",
       " 'chr1:14350': 'downstream_gene_variant',\n",
       " 'chr1:14354': 'downstream_gene_variant',\n",
       " 'chr1:14464': 'downstream_gene_variant',\n",
       " 'chr1:14469': 'downstream_gene_variant',\n",
       " 'chr1:14470': 'downstream_gene_variant',\n",
       " 'chr1:14487': 'downstream_gene_variant',\n",
       " 'chr1:14521': 'downstream_gene_variant',\n",
       " 'chr1:14522': 'downstream_gene_variant',\n",
       " 'chr1:14542': 'downstream_gene_variant',\n",
       " 'chr1:14574': 'downstream_gene_variant',\n",
       " 'chr1:14584': 'downstream_gene_variant',\n",
       " 'chr1:14590': 'downstream_gene_variant',\n",
       " 'chr1:14599': 'downstream_gene_variant',\n",
       " 'chr1:14604': 'downstream_gene_variant',\n",
       " 'chr1:14610': 'downstream_gene_variant',\n",
       " 'chr1:14626': 'downstream_gene_variant',\n",
       " 'chr1:14643': 'downstream_gene_variant',\n",
       " 'chr1:14653': 'downstream_gene_variant',\n",
       " 'chr1:14671': 'downstream_gene_variant',\n",
       " 'chr1:14673': 'downstream_gene_variant',\n",
       " 'chr1:14677': 'downstream_gene_variant',\n",
       " 'chr1:14699': 'downstream_gene_variant',\n",
       " 'chr1:14711': 'downstream_gene_variant',\n",
       " 'chr1:14719': 'downstream_gene_variant',\n",
       " 'chr1:14752': 'downstream_gene_variant',\n",
       " 'chr1:14763': 'downstream_gene_variant',\n",
       " 'chr1:14773': 'downstream_gene_variant',\n",
       " 'chr1:14775': 'downstream_gene_variant',\n",
       " 'chr1:14792': 'downstream_gene_variant',\n",
       " 'chr1:14842': 'downstream_gene_variant',\n",
       " 'chr1:14843': 'downstream_gene_variant',\n",
       " 'chr1:14852': 'downstream_gene_variant',\n",
       " 'chr1:14860': 'downstream_gene_variant',\n",
       " 'chr1:14896': 'downstream_gene_variant',\n",
       " 'chr1:14906': 'downstream_gene_variant',\n",
       " 'chr1:14907': 'downstream_gene_variant',\n",
       " 'chr1:14930': 'downstream_gene_variant',\n",
       " 'chr1:14932': 'downstream_gene_variant',\n",
       " 'chr1:14933': 'downstream_gene_variant',\n",
       " 'chr1:14937': 'downstream_gene_variant',\n",
       " 'chr1:14948': 'downstream_gene_variant',\n",
       " 'chr1:14975': 'downstream_gene_variant',\n",
       " 'chr1:14976': 'downstream_gene_variant',\n",
       " 'chr1:15015': 'downstream_gene_variant',\n",
       " 'chr1:15045': 'downstream_gene_variant',\n",
       " 'chr1:15052': 'downstream_gene_variant',\n",
       " 'chr1:15094': 'downstream_gene_variant',\n",
       " 'chr1:15112': 'downstream_gene_variant',\n",
       " 'chr1:15118': 'downstream_gene_variant',\n",
       " 'chr1:15189': 'downstream_gene_variant',\n",
       " 'chr1:15207': 'downstream_gene_variant',\n",
       " 'chr1:15208': 'downstream_gene_variant',\n",
       " 'chr1:15211': 'downstream_gene_variant',\n",
       " 'chr1:15259': 'downstream_gene_variant',\n",
       " 'chr1:15273': 'downstream_gene_variant',\n",
       " 'chr1:15274': 'downstream_gene_variant',\n",
       " 'chr1:15351': 'downstream_gene_variant',\n",
       " 'chr1:15352': 'downstream_gene_variant',\n",
       " 'chr1:15354': 'downstream_gene_variant',\n",
       " 'chr1:15362': 'downstream_gene_variant',\n",
       " 'chr1:15364': 'downstream_gene_variant',\n",
       " 'chr1:15374': 'downstream_gene_variant',\n",
       " 'chr1:15381': 'downstream_gene_variant',\n",
       " 'chr1:15390': 'downstream_gene_variant',\n",
       " 'chr1:15446': 'downstream_gene_variant',\n",
       " 'chr1:15447': 'downstream_gene_variant',\n",
       " 'chr1:15484': 'downstream_gene_variant',\n",
       " 'chr1:15514': 'downstream_gene_variant',\n",
       " 'chr1:15515': 'downstream_gene_variant',\n",
       " 'chr1:15521': 'downstream_gene_variant',\n",
       " 'chr1:15527': 'downstream_gene_variant',\n",
       " 'chr1:15566': 'downstream_gene_variant',\n",
       " 'chr1:15570': 'downstream_gene_variant',\n",
       " 'chr1:15584': 'downstream_gene_variant',\n",
       " 'chr1:15585': 'downstream_gene_variant',\n",
       " 'chr1:15589': 'downstream_gene_variant',\n",
       " 'chr1:15616': 'downstream_gene_variant',\n",
       " 'chr1:15634': 'downstream_gene_variant',\n",
       " 'chr1:15644': 'downstream_gene_variant',\n",
       " 'chr1:15647': 'downstream_gene_variant',\n",
       " 'chr1:15694': 'downstream_gene_variant',\n",
       " 'chr1:15715': 'downstream_gene_variant',\n",
       " 'chr1:15717': 'downstream_gene_variant',\n",
       " 'chr1:15721': 'downstream_gene_variant',\n",
       " 'chr1:15770': 'downstream_gene_variant',\n",
       " 'chr1:15773': 'downstream_gene_variant',\n",
       " 'chr1:15774': 'downstream_gene_variant',\n",
       " 'chr1:15776': 'downstream_gene_variant',\n",
       " 'chr1:15777': 'downstream_gene_variant',\n",
       " 'chr1:15790': 'downstream_gene_variant',\n",
       " 'chr1:15811': 'downstream_gene_variant',\n",
       " 'chr1:15820': 'downstream_gene_variant',\n",
       " 'chr1:15835': 'downstream_gene_variant',\n",
       " 'chr1:15849': 'downstream_gene_variant',\n",
       " 'chr1:15906': 'downstream_gene_variant',\n",
       " 'chr1:16014': 'downstream_gene_variant',\n",
       " 'chr1:16062': 'downstream_gene_variant',\n",
       " 'chr1:16068': 'downstream_gene_variant',\n",
       " 'chr1:16103': 'downstream_gene_variant',\n",
       " 'chr1:16111': 'downstream_gene_variant',\n",
       " 'chr1:16125': 'downstream_gene_variant',\n",
       " 'chr1:16140': 'downstream_gene_variant',\n",
       " 'chr1:16141': 'downstream_gene_variant',\n",
       " 'chr1:16229': 'downstream_gene_variant',\n",
       " 'chr1:16243': 'downstream_gene_variant',\n",
       " 'chr1:16257': 'downstream_gene_variant',\n",
       " 'chr1:16263': 'downstream_gene_variant',\n",
       " 'chr1:16280': 'downstream_gene_variant',\n",
       " 'chr1:16288': 'downstream_gene_variant',\n",
       " 'chr1:16298': 'downstream_gene_variant',\n",
       " 'chr1:16350': 'downstream_gene_variant',\n",
       " 'chr1:16378': 'downstream_gene_variant',\n",
       " 'chr1:16433': 'downstream_gene_variant',\n",
       " 'chr1:16451': 'downstream_gene_variant',\n",
       " 'chr1:16474': 'downstream_gene_variant',\n",
       " 'chr1:16487': 'downstream_gene_variant',\n",
       " 'chr1:16495': 'downstream_gene_variant',\n",
       " 'chr1:16534': 'downstream_gene_variant',\n",
       " 'chr1:16555': 'downstream_gene_variant',\n",
       " 'chr1:16567': 'downstream_gene_variant',\n",
       " 'chr1:16570': 'downstream_gene_variant',\n",
       " 'chr1:16571': 'downstream_gene_variant',\n",
       " 'chr1:16576': 'downstream_gene_variant',\n",
       " 'chr1:16617': 'downstream_gene_variant',\n",
       " 'chr1:16619': 'downstream_gene_variant',\n",
       " 'chr1:16631': 'downstream_gene_variant',\n",
       " 'chr1:16647': 'downstream_gene_variant',\n",
       " 'chr1:16682': 'downstream_gene_variant',\n",
       " 'chr1:16686': 'downstream_gene_variant',\n",
       " 'chr1:16688': 'downstream_gene_variant',\n",
       " 'chr1:16689': 'downstream_gene_variant',\n",
       " 'chr1:16703': 'downstream_gene_variant',\n",
       " 'chr1:16711': 'downstream_gene_variant',\n",
       " 'chr1:16713': 'downstream_gene_variant',\n",
       " 'chr1:16719': 'downstream_gene_variant',\n",
       " 'chr1:16725': 'downstream_gene_variant',\n",
       " 'chr1:16731': 'downstream_gene_variant',\n",
       " 'chr1:16734': 'downstream_gene_variant',\n",
       " 'chr1:16737': 'downstream_gene_variant',\n",
       " 'chr1:16742': 'downstream_gene_variant',\n",
       " 'chr1:16792': 'downstream_gene_variant',\n",
       " 'chr1:16798': 'downstream_gene_variant',\n",
       " 'chr1:16805': 'downstream_gene_variant',\n",
       " 'chr1:16806': 'downstream_gene_variant',\n",
       " 'chr1:16809': 'downstream_gene_variant',\n",
       " 'chr1:16825': 'downstream_gene_variant',\n",
       " 'chr1:16830': 'downstream_gene_variant',\n",
       " 'chr1:16841': 'downstream_gene_variant',\n",
       " 'chr1:16850': 'downstream_gene_variant',\n",
       " 'chr1:16856': 'downstream_gene_variant',\n",
       " 'chr1:16888': 'downstream_gene_variant',\n",
       " 'chr1:16933': 'downstream_gene_variant',\n",
       " 'chr1:16936': 'downstream_gene_variant',\n",
       " 'chr1:16946': 'downstream_gene_variant',\n",
       " 'chr1:16949': 'downstream_gene_variant',\n",
       " 'chr1:16957': 'downstream_gene_variant',\n",
       " 'chr1:16963': 'downstream_gene_variant',\n",
       " 'chr1:16968': 'downstream_gene_variant',\n",
       " 'chr1:16975': 'downstream_gene_variant',\n",
       " 'chr1:16977': 'downstream_gene_variant',\n",
       " 'chr1:16993': 'downstream_gene_variant',\n",
       " 'chr1:16996': 'downstream_gene_variant',\n",
       " 'chr1:17005': 'downstream_gene_variant',\n",
       " 'chr1:17006': 'downstream_gene_variant',\n",
       " 'chr1:17020': 'downstream_gene_variant',\n",
       " 'chr1:17036': 'downstream_gene_variant',\n",
       " 'chr1:17038': 'downstream_gene_variant',\n",
       " 'chr1:17041': 'downstream_gene_variant',\n",
       " 'chr1:17046': 'downstream_gene_variant',\n",
       " 'chr1:17101': 'downstream_gene_variant',\n",
       " 'chr1:17147': 'downstream_gene_variant',\n",
       " 'chr1:17156': 'downstream_gene_variant',\n",
       " 'chr1:17172': 'downstream_gene_variant',\n",
       " 'chr1:17181': 'downstream_gene_variant',\n",
       " 'chr1:17192': 'downstream_gene_variant',\n",
       " 'chr1:17205': 'downstream_gene_variant',\n",
       " 'chr1:17207': 'downstream_gene_variant',\n",
       " 'chr1:17212': 'downstream_gene_variant',\n",
       " 'chr1:17213': 'downstream_gene_variant',\n",
       " 'chr1:17222': 'downstream_gene_variant',\n",
       " 'chr1:17223': 'downstream_gene_variant',\n",
       " 'chr1:17294': 'downstream_gene_variant',\n",
       " 'chr1:17365': 'downstream_gene_variant',\n",
       " 'chr1:17375': 'mature_miRNA_variant',\n",
       " 'chr1:17378': 'mature_miRNA_variant',\n",
       " 'chr1:17379': 'mature_miRNA_variant',\n",
       " 'chr1:17385': 'mature_miRNA_variant',\n",
       " 'chr1:17398': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:17403': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:17404': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:17406': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:17407': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:17408': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:17452': 'upstream_gene_variant',\n",
       " 'chr1:17472': 'upstream_gene_variant',\n",
       " 'chr1:17478': 'upstream_gene_variant',\n",
       " 'chr1:17479': 'upstream_gene_variant',\n",
       " 'chr1:17484': 'upstream_gene_variant',\n",
       " 'chr1:17492': 'upstream_gene_variant',\n",
       " 'chr1:17493': 'upstream_gene_variant',\n",
       " 'chr1:17501': 'upstream_gene_variant',\n",
       " 'chr1:17512': 'upstream_gene_variant',\n",
       " 'chr1:17519': 'upstream_gene_variant',\n",
       " 'chr1:17538': 'upstream_gene_variant',\n",
       " 'chr1:17551': 'upstream_gene_variant',\n",
       " 'chr1:17556': 'upstream_gene_variant',\n",
       " 'chr1:17559': 'upstream_gene_variant',\n",
       " 'chr1:17562': 'upstream_gene_variant',\n",
       " 'chr1:17564': 'upstream_gene_variant',\n",
       " 'chr1:17589': 'upstream_gene_variant',\n",
       " 'chr1:17592': 'upstream_gene_variant',\n",
       " 'chr1:17594': 'upstream_gene_variant',\n",
       " 'chr1:17606': 'upstream_gene_variant',\n",
       " 'chr1:17614': 'upstream_gene_variant',\n",
       " 'chr1:17626': 'upstream_gene_variant',\n",
       " 'chr1:17688': 'upstream_gene_variant',\n",
       " 'chr1:17689': 'upstream_gene_variant',\n",
       " 'chr1:17694': 'upstream_gene_variant',\n",
       " 'chr1:17695': 'upstream_gene_variant',\n",
       " 'chr1:17697': 'upstream_gene_variant',\n",
       " 'chr1:17701': 'upstream_gene_variant',\n",
       " 'chr1:17709': 'upstream_gene_variant',\n",
       " 'chr1:17716': 'upstream_gene_variant',\n",
       " 'chr1:17722': 'upstream_gene_variant',\n",
       " 'chr1:17728': 'upstream_gene_variant',\n",
       " 'chr1:17729': 'upstream_gene_variant',\n",
       " 'chr1:17730': 'upstream_gene_variant',\n",
       " 'chr1:17746': 'upstream_gene_variant',\n",
       " 'chr1:17750': 'upstream_gene_variant',\n",
       " 'chr1:17753': 'upstream_gene_variant',\n",
       " 'chr1:17765': 'upstream_gene_variant',\n",
       " 'chr1:17767': 'upstream_gene_variant',\n",
       " 'chr1:17781': 'upstream_gene_variant',\n",
       " 'chr1:17805': 'upstream_gene_variant',\n",
       " 'chr1:17829': 'upstream_gene_variant',\n",
       " 'chr1:17902': 'upstream_gene_variant',\n",
       " 'chr1:17928': 'upstream_gene_variant',\n",
       " 'chr1:17929': 'upstream_gene_variant',\n",
       " 'chr1:18164': 'upstream_gene_variant',\n",
       " 'chr1:18643': 'upstream_gene_variant',\n",
       " 'chr1:18746': 'upstream_gene_variant',\n",
       " 'chr1:18927': 'upstream_gene_variant',\n",
       " 'chr1:18934': 'upstream_gene_variant',\n",
       " 'chr1:18955': 'upstream_gene_variant',\n",
       " 'chr1:18996': 'upstream_gene_variant',\n",
       " 'chr1:19004': 'upstream_gene_variant',\n",
       " 'chr1:19015': 'upstream_gene_variant',\n",
       " 'chr1:19016': 'upstream_gene_variant',\n",
       " 'chr1:19044': 'upstream_gene_variant',\n",
       " 'chr1:19057': 'upstream_gene_variant',\n",
       " 'chr1:19090': 'upstream_gene_variant',\n",
       " 'chr1:19129': 'upstream_gene_variant',\n",
       " 'chr1:19133': 'upstream_gene_variant',\n",
       " 'chr1:19145': 'upstream_gene_variant',\n",
       " 'chr1:19172': 'upstream_gene_variant',\n",
       " 'chr1:19189': 'upstream_gene_variant',\n",
       " 'chr1:19194': 'upstream_gene_variant',\n",
       " 'chr1:19217': 'upstream_gene_variant',\n",
       " 'chr1:19226': 'upstream_gene_variant',\n",
       " 'chr1:19236': 'upstream_gene_variant',\n",
       " 'chr1:19248': 'upstream_gene_variant',\n",
       " 'chr1:19249': 'upstream_gene_variant',\n",
       " 'chr1:19254': 'upstream_gene_variant',\n",
       " 'chr1:19269': 'upstream_gene_variant',\n",
       " 'chr1:19279': 'upstream_gene_variant',\n",
       " 'chr1:19280': 'upstream_gene_variant',\n",
       " 'chr1:19296': 'upstream_gene_variant',\n",
       " 'chr1:19306': 'upstream_gene_variant',\n",
       " 'chr1:19321': 'upstream_gene_variant',\n",
       " 'chr1:19322': 'upstream_gene_variant',\n",
       " 'chr1:19342': 'upstream_gene_variant',\n",
       " 'chr1:19355': 'upstream_gene_variant',\n",
       " 'chr1:19365': 'upstream_gene_variant',\n",
       " 'chr1:19374': 'upstream_gene_variant',\n",
       " 'chr1:19391': 'upstream_gene_variant',\n",
       " 'chr1:19392': 'upstream_gene_variant',\n",
       " 'chr1:19402': 'upstream_gene_variant',\n",
       " 'chr1:19419': 'upstream_gene_variant',\n",
       " 'chr1:19425': 'upstream_gene_variant',\n",
       " 'chr1:19432': 'upstream_gene_variant',\n",
       " 'chr1:19502': 'upstream_gene_variant',\n",
       " 'chr1:19506': 'upstream_gene_variant',\n",
       " 'chr1:19508': 'upstream_gene_variant',\n",
       " 'chr1:19509': 'upstream_gene_variant',\n",
       " 'chr1:19548': 'upstream_gene_variant',\n",
       " 'chr1:19555': 'upstream_gene_variant',\n",
       " 'chr1:19583': 'upstream_gene_variant',\n",
       " 'chr1:19589': 'upstream_gene_variant',\n",
       " 'chr1:19749': 'upstream_gene_variant',\n",
       " 'chr1:19765': 'upstream_gene_variant',\n",
       " 'chr1:19789': 'upstream_gene_variant',\n",
       " 'chr1:19858': 'upstream_gene_variant',\n",
       " 'chr1:19889': 'upstream_gene_variant',\n",
       " 'chr1:19918': 'upstream_gene_variant',\n",
       " 'chr1:19919': 'upstream_gene_variant',\n",
       " 'chr1:19944': 'upstream_gene_variant',\n",
       " 'chr1:19952': 'upstream_gene_variant',\n",
       " 'chr1:19980': 'upstream_gene_variant',\n",
       " 'chr1:20037': 'upstream_gene_variant',\n",
       " 'chr1:20044': 'upstream_gene_variant',\n",
       " 'chr1:20045': 'upstream_gene_variant',\n",
       " 'chr1:20104': 'upstream_gene_variant',\n",
       " 'chr1:20128': 'upstream_gene_variant',\n",
       " 'chr1:20129': 'upstream_gene_variant',\n",
       " 'chr1:20144': 'upstream_gene_variant',\n",
       " 'chr1:20158': 'upstream_gene_variant',\n",
       " 'chr1:20166': 'upstream_gene_variant',\n",
       " 'chr1:20184': 'upstream_gene_variant',\n",
       " 'chr1:20185': 'upstream_gene_variant',\n",
       " 'chr1:20191': 'upstream_gene_variant',\n",
       " 'chr1:20206': 'upstream_gene_variant',\n",
       " 'chr1:20207': 'upstream_gene_variant',\n",
       " 'chr1:20212': 'upstream_gene_variant',\n",
       " 'chr1:20217': 'upstream_gene_variant',\n",
       " 'chr1:20225': 'upstream_gene_variant',\n",
       " 'chr1:20227': 'upstream_gene_variant',\n",
       " 'chr1:20231': 'upstream_gene_variant',\n",
       " 'chr1:20235': 'upstream_gene_variant',\n",
       " 'chr1:20240': 'upstream_gene_variant',\n",
       " 'chr1:20245': 'upstream_gene_variant',\n",
       " 'chr1:20250': 'upstream_gene_variant',\n",
       " 'chr1:20254': 'upstream_gene_variant',\n",
       " 'chr1:20309': 'upstream_gene_variant',\n",
       " 'chr1:20321': 'upstream_gene_variant',\n",
       " 'chr1:20343': 'upstream_gene_variant',\n",
       " 'chr1:20357': 'upstream_gene_variant',\n",
       " 'chr1:20388': 'upstream_gene_variant',\n",
       " 'chr1:20413': 'upstream_gene_variant',\n",
       " 'chr1:20443': 'upstream_gene_variant',\n",
       " 'chr1:20481': 'upstream_gene_variant',\n",
       " 'chr1:20504': 'upstream_gene_variant',\n",
       " 'chr1:20507': 'upstream_gene_variant',\n",
       " 'chr1:20522': 'upstream_gene_variant',\n",
       " 'chr1:20547': 'upstream_gene_variant',\n",
       " 'chr1:20552': 'upstream_gene_variant',\n",
       " 'chr1:20595': 'upstream_gene_variant',\n",
       " 'chr1:20600': 'upstream_gene_variant',\n",
       " 'chr1:20605': 'upstream_gene_variant',\n",
       " 'chr1:20620': 'upstream_gene_variant',\n",
       " 'chr1:20627': 'upstream_gene_variant',\n",
       " 'chr1:20628': 'upstream_gene_variant',\n",
       " 'chr1:20630': 'upstream_gene_variant',\n",
       " 'chr1:21224': 'upstream_gene_variant',\n",
       " 'chr1:21319': 'upstream_gene_variant',\n",
       " 'chr1:21411': 'upstream_gene_variant',\n",
       " 'chr1:21420': 'upstream_gene_variant',\n",
       " 'chr1:21489': 'upstream_gene_variant',\n",
       " 'chr1:21600': 'upstream_gene_variant',\n",
       " 'chr1:21674': 'upstream_gene_variant',\n",
       " 'chr1:21715': 'upstream_gene_variant',\n",
       " 'chr1:21769': 'upstream_gene_variant',\n",
       " 'chr1:21796': 'upstream_gene_variant',\n",
       " 'chr1:21801': 'upstream_gene_variant',\n",
       " 'chr1:21812': 'upstream_gene_variant',\n",
       " 'chr1:23020': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23067': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23071': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23097': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23105': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23108': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23115': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23125': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23134': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23149': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23151': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23200': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23255': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23343': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23354': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23357': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23358': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23458': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23506': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23507': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23538': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23559': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23584': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23647': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:23653': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24511': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24534': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24592': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24638': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24648': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24898': 'splice_region_variant,intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24912': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:24924': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:28198': 'upstream_gene_variant',\n",
       " 'chr1:28214': 'upstream_gene_variant',\n",
       " 'chr1:28251': 'upstream_gene_variant',\n",
       " 'chr1:28253': 'upstream_gene_variant',\n",
       " 'chr1:28254': 'upstream_gene_variant',\n",
       " 'chr1:28343': 'upstream_gene_variant',\n",
       " 'chr1:28354': 'upstream_gene_variant',\n",
       " 'chr1:28367': 'upstream_gene_variant',\n",
       " 'chr1:28376': 'upstream_gene_variant',\n",
       " 'chr1:28480': 'upstream_gene_variant',\n",
       " 'chr1:28494': 'upstream_gene_variant',\n",
       " 'chr1:28511': 'upstream_gene_variant',\n",
       " 'chr1:28516': 'upstream_gene_variant',\n",
       " 'chr1:28558': 'upstream_gene_variant',\n",
       " 'chr1:28563': 'upstream_gene_variant',\n",
       " 'chr1:28578': 'upstream_gene_variant',\n",
       " 'chr1:28592': 'upstream_gene_variant',\n",
       " 'chr1:28593': 'upstream_gene_variant',\n",
       " 'chr1:28608': 'upstream_gene_variant',\n",
       " 'chr1:28663': 'upstream_gene_variant',\n",
       " 'chr1:28682': 'upstream_gene_variant',\n",
       " 'chr1:28692': 'upstream_gene_variant',\n",
       " 'chr1:28693': 'upstream_gene_variant',\n",
       " 'chr1:28815': 'upstream_gene_variant',\n",
       " 'chr1:28826': 'upstream_gene_variant',\n",
       " 'chr1:28835': 'upstream_gene_variant',\n",
       " 'chr1:28863': 'upstream_gene_variant',\n",
       " 'chr1:28875': 'upstream_gene_variant',\n",
       " 'chr1:28905': 'upstream_gene_variant',\n",
       " 'chr1:28957': 'upstream_gene_variant',\n",
       " 'chr1:28959': 'upstream_gene_variant',\n",
       " 'chr1:28976': 'upstream_gene_variant',\n",
       " 'chr1:28977': 'upstream_gene_variant',\n",
       " 'chr1:29022': 'upstream_gene_variant',\n",
       " 'chr1:29100': 'upstream_gene_variant',\n",
       " 'chr1:29112': 'upstream_gene_variant',\n",
       " 'chr1:29138': 'upstream_gene_variant',\n",
       " 'chr1:29256': 'upstream_gene_variant',\n",
       " 'chr1:29262': 'upstream_gene_variant',\n",
       " 'chr1:29265': 'upstream_gene_variant',\n",
       " 'chr1:29280': 'upstream_gene_variant',\n",
       " 'chr1:29306': 'upstream_gene_variant',\n",
       " 'chr1:29362': 'upstream_gene_variant',\n",
       " 'chr1:29368': 'upstream_gene_variant',\n",
       " 'chr1:29443': 'upstream_gene_variant',\n",
       " 'chr1:29571': 'upstream_gene_variant',\n",
       " 'chr1:29575': 'upstream_gene_variant',\n",
       " 'chr1:29604': 'upstream_gene_variant',\n",
       " 'chr1:29616': 'upstream_gene_variant',\n",
       " 'chr1:29656': 'upstream_gene_variant',\n",
       " 'chr1:29757': 'upstream_gene_variant',\n",
       " 'chr1:29759': 'upstream_gene_variant',\n",
       " 'chr1:30524': 'downstream_gene_variant',\n",
       " 'chr1:30548': 'downstream_gene_variant',\n",
       " 'chr1:30570': 'downstream_gene_variant',\n",
       " 'chr1:30663': 'downstream_gene_variant',\n",
       " 'chr1:30672': 'downstream_gene_variant',\n",
       " 'chr1:30779': 'downstream_gene_variant',\n",
       " 'chr1:30794': 'downstream_gene_variant',\n",
       " 'chr1:30809': 'downstream_gene_variant',\n",
       " 'chr1:30855': 'downstream_gene_variant',\n",
       " 'chr1:30860': 'downstream_gene_variant',\n",
       " 'chr1:30885': 'downstream_gene_variant',\n",
       " 'chr1:30923': 'downstream_gene_variant',\n",
       " 'chr1:30935': 'downstream_gene_variant',\n",
       " 'chr1:31029': 'downstream_gene_variant',\n",
       " 'chr1:31058': 'downstream_gene_variant',\n",
       " 'chr1:31386': 'downstream_gene_variant',\n",
       " 'chr1:31395': 'downstream_gene_variant',\n",
       " 'chr1:31488': 'downstream_gene_variant',\n",
       " 'chr1:33054': 'downstream_gene_variant',\n",
       " 'chr1:33432': 'downstream_gene_variant',\n",
       " 'chr1:33448': 'downstream_gene_variant',\n",
       " 'chr1:33449': 'downstream_gene_variant',\n",
       " 'chr1:33521': 'downstream_gene_variant',\n",
       " 'chr1:33539': 'downstream_gene_variant',\n",
       " 'chr1:38606': 'upstream_gene_variant',\n",
       " 'chr1:38742': 'upstream_gene_variant',\n",
       " 'chr1:38775': 'upstream_gene_variant',\n",
       " 'chr1:38786': 'upstream_gene_variant',\n",
       " 'chr1:38800': 'upstream_gene_variant',\n",
       " 'chr1:38824': 'upstream_gene_variant',\n",
       " 'chr1:38854': 'upstream_gene_variant',\n",
       " 'chr1:38889': 'upstream_gene_variant',\n",
       " 'chr1:39015': 'upstream_gene_variant',\n",
       " 'chr1:39070': 'upstream_gene_variant',\n",
       " 'chr1:39101': 'upstream_gene_variant',\n",
       " 'chr1:39143': 'upstream_gene_variant',\n",
       " 'chr1:39167': 'upstream_gene_variant',\n",
       " 'chr1:39180': 'upstream_gene_variant',\n",
       " 'chr1:39185': 'upstream_gene_variant',\n",
       " 'chr1:39203': 'upstream_gene_variant',\n",
       " 'chr1:39220': 'upstream_gene_variant',\n",
       " 'chr1:39224': 'upstream_gene_variant',\n",
       " 'chr1:39230': 'upstream_gene_variant',\n",
       " 'chr1:39240': 'upstream_gene_variant',\n",
       " 'chr1:39247': 'upstream_gene_variant',\n",
       " 'chr1:39255': 'upstream_gene_variant',\n",
       " 'chr1:39256': 'upstream_gene_variant',\n",
       " 'chr1:39257': 'upstream_gene_variant',\n",
       " 'chr1:39261': 'upstream_gene_variant',\n",
       " 'chr1:39292': 'upstream_gene_variant',\n",
       " 'chr1:39362': 'upstream_gene_variant',\n",
       " 'chr1:39368': 'upstream_gene_variant',\n",
       " 'chr1:39379': 'upstream_gene_variant',\n",
       " 'chr1:39406': 'upstream_gene_variant',\n",
       " 'chr1:39423': 'upstream_gene_variant',\n",
       " 'chr1:39430': 'upstream_gene_variant',\n",
       " 'chr1:39501': 'upstream_gene_variant',\n",
       " 'chr1:39503': 'upstream_gene_variant',\n",
       " 'chr1:39530': 'upstream_gene_variant',\n",
       " 'chr1:39532': 'upstream_gene_variant',\n",
       " 'chr1:39578': 'upstream_gene_variant',\n",
       " 'chr1:39579': 'upstream_gene_variant',\n",
       " 'chr1:39602': 'upstream_gene_variant',\n",
       " 'chr1:39604': 'upstream_gene_variant',\n",
       " 'chr1:39631': 'upstream_gene_variant',\n",
       " 'chr1:39641': 'upstream_gene_variant',\n",
       " 'chr1:39662': 'upstream_gene_variant',\n",
       " 'chr1:39675': 'upstream_gene_variant',\n",
       " 'chr1:39676': 'upstream_gene_variant',\n",
       " 'chr1:39722': 'upstream_gene_variant',\n",
       " 'chr1:39734': 'upstream_gene_variant',\n",
       " 'chr1:39779': 'upstream_gene_variant',\n",
       " 'chr1:39830': 'upstream_gene_variant',\n",
       " 'chr1:39860': 'upstream_gene_variant',\n",
       " 'chr1:39864': 'upstream_gene_variant',\n",
       " 'chr1:39879': 'upstream_gene_variant',\n",
       " 'chr1:39883': 'upstream_gene_variant',\n",
       " 'chr1:39892': 'upstream_gene_variant',\n",
       " 'chr1:39893': 'upstream_gene_variant',\n",
       " 'chr1:39966': 'upstream_gene_variant',\n",
       " 'chr1:40023': 'upstream_gene_variant',\n",
       " 'chr1:40036': 'upstream_gene_variant',\n",
       " 'chr1:40049': 'upstream_gene_variant',\n",
       " 'chr1:40110': 'upstream_gene_variant',\n",
       " 'chr1:40123': 'upstream_gene_variant',\n",
       " 'chr1:40130': 'upstream_gene_variant',\n",
       " 'chr1:40149': 'upstream_gene_variant',\n",
       " 'chr1:40160': 'upstream_gene_variant',\n",
       " 'chr1:40174': 'upstream_gene_variant',\n",
       " 'chr1:40176': 'upstream_gene_variant',\n",
       " 'chr1:40208': 'upstream_gene_variant',\n",
       " 'chr1:40387': 'upstream_gene_variant',\n",
       " 'chr1:40389': 'upstream_gene_variant',\n",
       " 'chr1:40400': 'upstream_gene_variant',\n",
       " 'chr1:40705': 'upstream_gene_variant',\n",
       " 'chr1:40706': 'upstream_gene_variant',\n",
       " 'chr1:40894': 'upstream_gene_variant',\n",
       " 'chr1:46320': 'intergenic_variant',\n",
       " 'chr1:46401': 'intergenic_variant',\n",
       " 'chr1:46404': 'intergenic_variant',\n",
       " 'chr1:46417': 'intergenic_variant',\n",
       " 'chr1:46571': 'intergenic_variant',\n",
       " 'chr1:46610': 'intergenic_variant',\n",
       " 'chr1:46633': 'intergenic_variant',\n",
       " 'chr1:46871': 'intergenic_variant',\n",
       " 'chr1:46873': 'intergenic_variant',\n",
       " 'chr1:46930': 'intergenic_variant',\n",
       " 'chr1:46998': 'intergenic_variant',\n",
       " 'chr1:47098': 'intergenic_variant',\n",
       " 'chr1:47159': 'intergenic_variant',\n",
       " 'chr1:47267': 'intergenic_variant',\n",
       " 'chr1:47310': 'intergenic_variant',\n",
       " 'chr1:47313': 'intergenic_variant',\n",
       " 'chr1:47314': 'intergenic_variant',\n",
       " 'chr1:47317': 'intergenic_variant',\n",
       " 'chr1:47362': 'intergenic_variant',\n",
       " 'chr1:47435': 'intergenic_variant',\n",
       " 'chr1:47451': 'intergenic_variant',\n",
       " 'chr1:47463': 'intergenic_variant',\n",
       " 'chr1:47941': 'upstream_gene_variant',\n",
       " 'chr1:47960': 'upstream_gene_variant',\n",
       " 'chr1:47978': 'upstream_gene_variant',\n",
       " 'chr1:48067': 'upstream_gene_variant',\n",
       " 'chr1:48104': 'upstream_gene_variant',\n",
       " 'chr1:48127': 'upstream_gene_variant',\n",
       " 'chr1:48152': 'upstream_gene_variant',\n",
       " 'chr1:48185': 'upstream_gene_variant',\n",
       " 'chr1:48186': 'upstream_gene_variant',\n",
       " 'chr1:48264': 'upstream_gene_variant',\n",
       " 'chr1:48306': 'upstream_gene_variant',\n",
       " 'chr1:48327': 'upstream_gene_variant',\n",
       " 'chr1:48328': 'upstream_gene_variant',\n",
       " 'chr1:48353': 'upstream_gene_variant',\n",
       " 'chr1:48448': 'upstream_gene_variant',\n",
       " 'chr1:48466': 'upstream_gene_variant',\n",
       " 'chr1:48518': 'upstream_gene_variant',\n",
       " 'chr1:48979': 'upstream_gene_variant',\n",
       " 'chr1:49046': 'upstream_gene_variant',\n",
       " 'chr1:49051': 'upstream_gene_variant',\n",
       " 'chr1:49148': 'upstream_gene_variant',\n",
       " 'chr1:49157': 'upstream_gene_variant',\n",
       " 'chr1:49243': 'upstream_gene_variant',\n",
       " 'chr1:49272': 'upstream_gene_variant',\n",
       " 'chr1:49291': 'upstream_gene_variant',\n",
       " 'chr1:49298': 'upstream_gene_variant',\n",
       " 'chr1:49314': 'upstream_gene_variant',\n",
       " 'chr1:49318': 'upstream_gene_variant',\n",
       " 'chr1:49343': 'upstream_gene_variant',\n",
       " 'chr1:49363': 'upstream_gene_variant',\n",
       " 'chr1:49404': 'upstream_gene_variant',\n",
       " 'chr1:49427': 'upstream_gene_variant',\n",
       " 'chr1:49448': 'upstream_gene_variant',\n",
       " 'chr1:49451': 'upstream_gene_variant',\n",
       " 'chr1:49482': 'upstream_gene_variant',\n",
       " 'chr1:49515': 'upstream_gene_variant',\n",
       " 'chr1:49554': 'upstream_gene_variant',\n",
       " 'chr1:49775': 'upstream_gene_variant',\n",
       " 'chr1:50170': 'upstream_gene_variant',\n",
       " 'chr1:50600': 'upstream_gene_variant',\n",
       " 'chr1:50879': 'upstream_gene_variant',\n",
       " 'chr1:50891': 'upstream_gene_variant',\n",
       " 'chr1:51146': 'upstream_gene_variant',\n",
       " 'chr1:51206': 'upstream_gene_variant',\n",
       " 'chr1:51269': 'upstream_gene_variant',\n",
       " 'chr1:51367': 'upstream_gene_variant',\n",
       " 'chr1:51427': 'upstream_gene_variant',\n",
       " 'chr1:51476': 'upstream_gene_variant',\n",
       " 'chr1:51479': 'upstream_gene_variant',\n",
       " 'chr1:51523': 'upstream_gene_variant',\n",
       " 'chr1:51565': 'upstream_gene_variant',\n",
       " 'chr1:51576': 'upstream_gene_variant',\n",
       " 'chr1:51620': 'upstream_gene_variant',\n",
       " 'chr1:51628': 'upstream_gene_variant',\n",
       " 'chr1:51641': 'upstream_gene_variant',\n",
       " 'chr1:51643': 'upstream_gene_variant',\n",
       " 'chr1:51668': 'upstream_gene_variant',\n",
       " 'chr1:51673': 'upstream_gene_variant',\n",
       " 'chr1:51675': 'upstream_gene_variant',\n",
       " 'chr1:51685': 'upstream_gene_variant',\n",
       " 'chr1:51687': 'upstream_gene_variant',\n",
       " 'chr1:51696': 'upstream_gene_variant',\n",
       " 'chr1:51719': 'upstream_gene_variant',\n",
       " 'chr1:51744': 'upstream_gene_variant',\n",
       " 'chr1:51745': 'upstream_gene_variant',\n",
       " 'chr1:51802': 'upstream_gene_variant',\n",
       " 'chr1:51803': 'upstream_gene_variant',\n",
       " 'chr1:51806': 'upstream_gene_variant',\n",
       " 'chr1:51839': 'upstream_gene_variant',\n",
       " 'chr1:51864': 'upstream_gene_variant',\n",
       " 'chr1:51898': 'upstream_gene_variant',\n",
       " 'chr1:51902': 'upstream_gene_variant',\n",
       " 'chr1:51928': 'upstream_gene_variant',\n",
       " 'chr1:51951': 'upstream_gene_variant',\n",
       " 'chr1:52058': 'upstream_gene_variant',\n",
       " 'chr1:52096': 'upstream_gene_variant',\n",
       " 'chr1:52105': 'upstream_gene_variant',\n",
       " 'chr1:52238': 'upstream_gene_variant',\n",
       " 'chr1:52253': 'upstream_gene_variant',\n",
       " 'chr1:52364': 'upstream_gene_variant',\n",
       " 'chr1:52402': 'upstream_gene_variant',\n",
       " 'chr1:52506': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:52607': 'upstream_gene_variant',\n",
       " 'chr1:52686': 'upstream_gene_variant',\n",
       " 'chr1:52702': 'upstream_gene_variant',\n",
       " 'chr1:52834': 'upstream_gene_variant',\n",
       " 'chr1:53082': 'upstream_gene_variant',\n",
       " 'chr1:53212': 'upstream_gene_variant',\n",
       " 'chr1:53266': 'upstream_gene_variant',\n",
       " 'chr1:53268': 'upstream_gene_variant',\n",
       " 'chr1:53313': 'upstream_gene_variant',\n",
       " 'chr1:53424': 'upstream_gene_variant',\n",
       " 'chr1:53498': 'upstream_gene_variant',\n",
       " 'chr1:53745': 'upstream_gene_variant',\n",
       " 'chr1:53810': 'upstream_gene_variant',\n",
       " 'chr1:53920': 'upstream_gene_variant',\n",
       " 'chr1:54001': 'upstream_gene_variant',\n",
       " 'chr1:54043': 'upstream_gene_variant',\n",
       " 'chr1:54170': 'upstream_gene_variant',\n",
       " 'chr1:54222': 'upstream_gene_variant',\n",
       " 'chr1:54303': 'upstream_gene_variant',\n",
       " 'chr1:54353': 'upstream_gene_variant',\n",
       " 'chr1:54366': 'upstream_gene_variant',\n",
       " 'chr1:54380': 'upstream_gene_variant',\n",
       " 'chr1:54421': 'upstream_gene_variant',\n",
       " 'chr1:54490': 'upstream_gene_variant',\n",
       " 'chr1:54493': 'upstream_gene_variant',\n",
       " 'chr1:54564': 'upstream_gene_variant',\n",
       " 'chr1:54566': 'upstream_gene_variant',\n",
       " 'chr1:54586': 'upstream_gene_variant',\n",
       " 'chr1:54625': 'upstream_gene_variant',\n",
       " 'chr1:54652': 'upstream_gene_variant',\n",
       " 'chr1:54676': 'upstream_gene_variant',\n",
       " 'chr1:54687': 'upstream_gene_variant',\n",
       " 'chr1:54703': 'upstream_gene_variant',\n",
       " 'chr1:54708': 'upstream_gene_variant',\n",
       " 'chr1:54716': 'upstream_gene_variant',\n",
       " 'chr1:54753': 'upstream_gene_variant',\n",
       " 'chr1:54757': 'upstream_gene_variant',\n",
       " 'chr1:54789': 'upstream_gene_variant',\n",
       " 'chr1:54798': 'upstream_gene_variant',\n",
       " 'chr1:54803': 'upstream_gene_variant',\n",
       " 'chr1:54814': 'upstream_gene_variant',\n",
       " 'chr1:54815': 'upstream_gene_variant',\n",
       " 'chr1:54826': 'upstream_gene_variant',\n",
       " 'chr1:54830': 'upstream_gene_variant',\n",
       " 'chr1:54844': 'upstream_gene_variant',\n",
       " 'chr1:54850': 'upstream_gene_variant',\n",
       " 'chr1:54901': 'upstream_gene_variant',\n",
       " 'chr1:54933': 'upstream_gene_variant',\n",
       " 'chr1:54945': 'upstream_gene_variant',\n",
       " 'chr1:54975': 'upstream_gene_variant',\n",
       " 'chr1:55054': 'upstream_gene_variant',\n",
       " 'chr1:55085': 'upstream_gene_variant',\n",
       " 'chr1:55131': 'upstream_gene_variant',\n",
       " 'chr1:55136': 'upstream_gene_variant',\n",
       " 'chr1:55164': 'upstream_gene_variant',\n",
       " 'chr1:55195': 'upstream_gene_variant',\n",
       " 'chr1:55207': 'upstream_gene_variant',\n",
       " 'chr1:55225': 'upstream_gene_variant',\n",
       " 'chr1:55231': 'upstream_gene_variant',\n",
       " 'chr1:55248': 'upstream_gene_variant',\n",
       " 'chr1:55299': 'upstream_gene_variant',\n",
       " 'chr1:55326': 'upstream_gene_variant',\n",
       " 'chr1:55330': 'upstream_gene_variant',\n",
       " 'chr1:55416': 'upstream_gene_variant',\n",
       " 'chr1:55427': 'upstream_gene_variant',\n",
       " 'chr1:55487': 'upstream_gene_variant',\n",
       " 'chr1:55545': 'upstream_gene_variant',\n",
       " 'chr1:55565': 'upstream_gene_variant',\n",
       " 'chr1:55580': 'upstream_gene_variant',\n",
       " 'chr1:55719': 'upstream_gene_variant',\n",
       " 'chr1:55759': 'upstream_gene_variant',\n",
       " 'chr1:55852': 'upstream_gene_variant',\n",
       " 'chr1:55926': 'upstream_gene_variant',\n",
       " 'chr1:56019': 'upstream_gene_variant',\n",
       " 'chr1:56070': 'upstream_gene_variant',\n",
       " 'chr1:56130': 'upstream_gene_variant',\n",
       " 'chr1:56348': 'upstream_gene_variant',\n",
       " 'chr1:56353': 'upstream_gene_variant',\n",
       " 'chr1:56381': 'upstream_gene_variant',\n",
       " 'chr1:56420': 'upstream_gene_variant',\n",
       " 'chr1:56445': 'upstream_gene_variant',\n",
       " 'chr1:56582': 'upstream_gene_variant',\n",
       " 'chr1:56638': 'upstream_gene_variant',\n",
       " 'chr1:56644': 'upstream_gene_variant',\n",
       " 'chr1:56679': 'upstream_gene_variant',\n",
       " 'chr1:56799': 'upstream_gene_variant',\n",
       " 'chr1:56815': 'upstream_gene_variant',\n",
       " 'chr1:56931': 'upstream_gene_variant',\n",
       " 'chr1:57095': 'upstream_gene_variant',\n",
       " 'chr1:57136': 'upstream_gene_variant',\n",
       " 'chr1:57158': 'upstream_gene_variant',\n",
       " 'chr1:57246': 'upstream_gene_variant',\n",
       " 'chr1:57262': 'upstream_gene_variant',\n",
       " 'chr1:57264': 'upstream_gene_variant',\n",
       " 'chr1:57292': 'upstream_gene_variant',\n",
       " 'chr1:57376': 'upstream_gene_variant',\n",
       " 'chr1:57380': 'upstream_gene_variant',\n",
       " 'chr1:57407': 'upstream_gene_variant',\n",
       " 'chr1:57463': 'upstream_gene_variant',\n",
       " 'chr1:57483': 'upstream_gene_variant',\n",
       " 'chr1:57708': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57772': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57773': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57856': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57878': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57896': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57952': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57958': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:57999': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58175': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58176': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58193': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58211': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58272': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58323': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58395': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58396': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58590': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58703': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:58737': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:58771': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:58786': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:58814': 'non_coding_transcript_exon_variant',\n",
       " 'chr1:58866': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58901': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:58989': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59040': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59051': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59089': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59108': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59114': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59120': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59121': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59129': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59193': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59232': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59276': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59288': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59315': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59400': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59498': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59615': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59895': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59924': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59941': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:59981': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60005': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60066': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60147': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60196': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60212': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60224': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60249': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60293': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60332': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60351': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60407': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60408': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60499': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60562': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60655': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60726': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60791': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60811': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60925': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:60950': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61021': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61045': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61099': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61115': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61208': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61219': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61301': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61335': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61362': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61397': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61442': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61462': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61464': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61499': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61543': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61578': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61579': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61583': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61681': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61706': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61743': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61745': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61761': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61843': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61848': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61851': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61881': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61920': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61923': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61947': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61987': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61989': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:61993': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62015': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62054': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62070': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62124': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62153': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62156': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62157': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62162': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62190': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62203': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62271': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62303': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62360': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62484': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62509': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62574': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62575': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62578': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62595': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62617': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62716': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62728': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62730': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62738': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62777': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:62783': 'intron_variant,non_coding_transcript_variant',\n",
       " 'chr1:63093': 'non_coding_transcript_exon_variant',\n",
       " ...}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conseq_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "conseq = []\n",
    "num_snps = len(snp_df)\n",
    "for index, row in snp_df.iterrows():\n",
    "    query = row[0].replace('_', ':')\n",
    "    conseq.append(conseq_dict.get(query, 'NA'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_df['conseq'] = conseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'missense_variant,NMD_transcript_variant'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conseq_dict['chr1:100133310']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>cpma</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>conseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_100007048</td>\n",
       "      <td>175.076413</td>\n",
       "      <td>0.722275</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_100110377</td>\n",
       "      <td>179.462116</td>\n",
       "      <td>0.718867</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_100133310</td>\n",
       "      <td>240.862862</td>\n",
       "      <td>0.676731</td>\n",
       "      <td>missense_variant,NMD_transcript_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_100133311</td>\n",
       "      <td>240.862862</td>\n",
       "      <td>0.676731</td>\n",
       "      <td>missense_variant,NMD_transcript_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_100206504</td>\n",
       "      <td>392.056652</td>\n",
       "      <td>0.593199</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31961</th>\n",
       "      <td>chr22_50606762</td>\n",
       "      <td>138.655291</td>\n",
       "      <td>0.776340</td>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31962</th>\n",
       "      <td>chr22_50625611</td>\n",
       "      <td>211.999535</td>\n",
       "      <td>0.725075</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31963</th>\n",
       "      <td>chr22_50626933</td>\n",
       "      <td>294.377515</td>\n",
       "      <td>0.678135</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31964</th>\n",
       "      <td>chr22_50679152</td>\n",
       "      <td>56.247171</td>\n",
       "      <td>0.856474</td>\n",
       "      <td>missense_variant,NMD_transcript_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31965</th>\n",
       "      <td>chr22_50744827</td>\n",
       "      <td>56.022188</td>\n",
       "      <td>0.856760</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31966 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp        cpma    pvalue  \\\n",
       "0      chr1_100007048  175.076413  0.722275   \n",
       "1      chr1_100110377  179.462116  0.718867   \n",
       "2      chr1_100133310  240.862862  0.676731   \n",
       "3      chr1_100133311  240.862862  0.676731   \n",
       "4      chr1_100206504  392.056652  0.593199   \n",
       "...               ...         ...       ...   \n",
       "31961  chr22_50606762  138.655291  0.776340   \n",
       "31962  chr22_50625611  211.999535  0.725075   \n",
       "31963  chr22_50626933  294.377515  0.678135   \n",
       "31964  chr22_50679152   56.247171  0.856474   \n",
       "31965  chr22_50744827   56.022188  0.856760   \n",
       "\n",
       "                                        conseq  \n",
       "0                               intron_variant  \n",
       "1           non_coding_transcript_exon_variant  \n",
       "2      missense_variant,NMD_transcript_variant  \n",
       "3      missense_variant,NMD_transcript_variant  \n",
       "4                             missense_variant  \n",
       "...                                        ...  \n",
       "31961                       synonymous_variant  \n",
       "31962                    upstream_gene_variant  \n",
       "31963                    upstream_gene_variant  \n",
       "31964  missense_variant,NMD_transcript_variant  \n",
       "31965                  downstream_gene_variant  \n",
       "\n",
       "[31966 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>cpma</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>conseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_100133310</td>\n",
       "      <td>240.862862</td>\n",
       "      <td>0.676731</td>\n",
       "      <td>missense_variant,NMD_transcript_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_100133311</td>\n",
       "      <td>240.862862</td>\n",
       "      <td>0.676731</td>\n",
       "      <td>missense_variant,NMD_transcript_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_100206504</td>\n",
       "      <td>392.056652</td>\n",
       "      <td>0.593199</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1_100484304</td>\n",
       "      <td>179.460522</td>\n",
       "      <td>0.718867</td>\n",
       "      <td>missense_variant,NMD_transcript_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1_100539534</td>\n",
       "      <td>2.509463</td>\n",
       "      <td>0.965886</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31935</th>\n",
       "      <td>chr22_50439767</td>\n",
       "      <td>1850.478687</td>\n",
       "      <td>0.290985</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31953</th>\n",
       "      <td>chr22_50547274</td>\n",
       "      <td>305.945676</td>\n",
       "      <td>0.672151</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31955</th>\n",
       "      <td>chr22_50548858</td>\n",
       "      <td>42.528216</td>\n",
       "      <td>0.875050</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31957</th>\n",
       "      <td>chr22_50549764</td>\n",
       "      <td>232.685136</td>\n",
       "      <td>0.712417</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31964</th>\n",
       "      <td>chr22_50679152</td>\n",
       "      <td>56.247171</td>\n",
       "      <td>0.856474</td>\n",
       "      <td>missense_variant,NMD_transcript_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6318 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp         cpma    pvalue  \\\n",
       "2      chr1_100133310   240.862862  0.676731   \n",
       "3      chr1_100133311   240.862862  0.676731   \n",
       "4      chr1_100206504   392.056652  0.593199   \n",
       "5      chr1_100484304   179.460522  0.718867   \n",
       "6      chr1_100539534     2.509463  0.965886   \n",
       "...               ...          ...       ...   \n",
       "31935  chr22_50439767  1850.478687  0.290985   \n",
       "31953  chr22_50547274   305.945676  0.672151   \n",
       "31955  chr22_50548858    42.528216  0.875050   \n",
       "31957  chr22_50549764   232.685136  0.712417   \n",
       "31964  chr22_50679152    56.247171  0.856474   \n",
       "\n",
       "                                        conseq  \n",
       "2      missense_variant,NMD_transcript_variant  \n",
       "3      missense_variant,NMD_transcript_variant  \n",
       "4                             missense_variant  \n",
       "5      missense_variant,NMD_transcript_variant  \n",
       "6                             missense_variant  \n",
       "...                                        ...  \n",
       "31935                         missense_variant  \n",
       "31953                         missense_variant  \n",
       "31955                         missense_variant  \n",
       "31957                         missense_variant  \n",
       "31964  missense_variant,NMD_transcript_variant  \n",
       "\n",
       "[6318 rows x 4 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df[snp_df['conseq'].str.contains('missense_variant')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>cpma</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>conseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_100206504</td>\n",
       "      <td>392.056652</td>\n",
       "      <td>0.593199</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1_100539534</td>\n",
       "      <td>2.509463</td>\n",
       "      <td>0.965886</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1_1014228</td>\n",
       "      <td>176.232881</td>\n",
       "      <td>0.721449</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1_1071842</td>\n",
       "      <td>402.678947</td>\n",
       "      <td>0.588111</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr1_1072052</td>\n",
       "      <td>111.911694</td>\n",
       "      <td>0.776392</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31910</th>\n",
       "      <td>chr22_50201536</td>\n",
       "      <td>1567.183664</td>\n",
       "      <td>0.332097</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31935</th>\n",
       "      <td>chr22_50439767</td>\n",
       "      <td>1850.478687</td>\n",
       "      <td>0.290985</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31953</th>\n",
       "      <td>chr22_50547274</td>\n",
       "      <td>305.945676</td>\n",
       "      <td>0.672151</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31955</th>\n",
       "      <td>chr22_50548858</td>\n",
       "      <td>42.528216</td>\n",
       "      <td>0.875050</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31957</th>\n",
       "      <td>chr22_50549764</td>\n",
       "      <td>232.685136</td>\n",
       "      <td>0.712417</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5712 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp         cpma    pvalue            conseq\n",
       "4      chr1_100206504   392.056652  0.593199  missense_variant\n",
       "6      chr1_100539534     2.509463  0.965886  missense_variant\n",
       "8        chr1_1014228   176.232881  0.721449  missense_variant\n",
       "36       chr1_1071842   402.678947  0.588111  missense_variant\n",
       "37       chr1_1072052   111.911694  0.776392  missense_variant\n",
       "...               ...          ...       ...               ...\n",
       "31910  chr22_50201536  1567.183664  0.332097  missense_variant\n",
       "31935  chr22_50439767  1850.478687  0.290985  missense_variant\n",
       "31953  chr22_50547274   305.945676  0.672151  missense_variant\n",
       "31955  chr22_50548858    42.528216  0.875050  missense_variant\n",
       "31957  chr22_50549764   232.685136  0.712417  missense_variant\n",
       "\n",
       "[5712 rows x 4 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df.loc[snp_df['conseq'] == 'missense_variant']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missense_a = empirical_conseq_a.loc[empirical_conseq_a['INFO'] == 'missense_variant']\n",
    "synonymous_a = empirical_conseq_a.loc[empirical_conseq_a['INFO'] == 'synonymous_variant']\n",
    "frameshift_a = empirical_conseq_a.loc[empirical_conseq_a['INFO'].isin(['frameshift_variant', 'frameshift_variant&NMD_transcript_variant', 'frameshift_variant&splice_region_variant','frameshift_variant&stop_lost','frameshift_variant&stop_retained_variant'])]\n",
    "stop_a = empirical_conseq_a.loc[empirical_conseq_a['INFO'].isin(['stop_gained', 'stop_gained&NMD_transcript_variant', 'stop_gained&splice_region_variant', 'stop_lost', 'stop_retained_variant'])]\n",
    "splice_a = empirical_conseq_a.loc[empirical_conseq_a['INFO'].isin(['splice_acceptor_variant','splice_acceptor_variant&coding_sequence_variant&intron_variant','splice_acceptor_variant&intron_variant','splice_donor_variant','splice_donor_variant&coding_sequence_variant&intron_variant','splice_donor_variant&intron_variant','splice_donor_variant&non_coding_transcript_variant','splice_region_variant&3_prime_UTR_variant','splice_region_variant&3_prime_UTR_variant&NMD_transcript_variant','splice_region_variant&5_prime_UTR_variant','splice_region_variant&intron_variant','splice_region_variant&intron_variant&NMD_transcript_variant','splice_region_variant&intron_variant&non_coding_transcript_variant','splice_region_variant&non_coding_transcript_exon_variant&non_coding_transcript_variant','splice_region_variant&synonymous_variant','splice_region_variant&synonymous_variant&NMD_transcript_variant'])]\n",
    "utr_a = empirical_conseq_a.loc[empirical_conseq_a['INFO'].isin(['3_prime_UTR_variant','3_prime_UTR_variant&NMD_transcript_variant','5_prime_UTR_variant','5_prime_UTR_variant&NMD_transcript_variant'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "value = conseq_dict.get('chr5:94386345', 'NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NA'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['intron_variant']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'intron_variant'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conseq_dict['chr1:100007048']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.5641619220421699e-06"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "0.05/len(snp_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2.259508713792433e-05]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pvalues_adj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    31966\n",
       "Name: pvalue, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(snp_df['pvalue'] <= (0.05/len(snp_df))).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    31810\n",
       "True       156\n",
       "Name: pvalue, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(snp_df['pvalue'] <= (0.05)).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (python3.6)",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
